### PR TITLE
Fix Pkg.free() so that it never removes the specified package

### PR DESCRIPTION
**Pull request by nalimilan**
The first commit adds a test which is supposed to fail on Windows only. The second commit fixes it.

However, I don't think this solves #17994, since the change just allows the package removal to complete without breaking further `Pkg` calls. We should probably also change `Pkg.free("X")` to add `X` to REQUIRE so that there's no risk of the package being removed by `resolve`, since this is what #17994 is about.

### CODE REVIEW

**Comment by nalimilan** [comment]
First commit failed on Windows with the expected error. Second commit passed the tests as it should. Third commit is a more radical version which essentially does the same thing, but removes duplicated methods.

The fourth commit attempts actually fixing #17994 by always adding the specified package to REQUIRE so that it's never removed. This sounds like a good idea since it's unlikely that the user intended to remove that package by calling `Pkg.free`. OTOH, any package not in REQUIRE can be removed (permanently, including possibly code in git branches) at any time by `resolve` when doing Pkg operations, so I'm not sure this semi-fix applying only to `Pkg.free` is enough.

**Comment by DrKrar** [comment]
But wouldn't put a package in REQUIRE will make it permanent even if the user accidentally install it because it is a requirement for some package he needs?

**Comment by nalimilan** [comment]
Yes. But the idea is that if you call `Pkg.free`, you did something special with that package anyway, so you can handle it manually. You should never have to call `Pkg.free` on a package you don't care about.

File: base/pkg/entry.jl [hunk 1]
```diff
@@ -250,11 +250,11 @@ function free(pkg::AbstractString)
             for ver in vers
                 sha1 = avail[ver].sha1
                 LibGit2.iscommit(sha1, repo) || continue
-                return LibGit2.transact(repo) do r
+                LibGit2.transact(repo) do r
                     LibGit2.isdirty(repo) && throw(PkgError("$pkg is dirty, bailing"))
                     LibGit2.checkout!(repo, sha1)
-                    resolve()
                 end
+                return resolve()
```
Comment (tkelman) [inline base/pkg/entry.jl]: this version isn't right either, as if the checkout fails you should restore the previous state, and not run resolve. likewise if resolve fails after checkout, should roll back.

Comment (nalimilan) [reply base/pkg/entry.jl]: I should have added a disclaimer that I blindly tried to fix the bug without being familiar with that code at all, so I wouldn't be surprised if that's totally incorrect. I can just note that the tests didn't cover that apparently, and that AFAICT the existing `free(pkgs)` method seems to work the same (but maybe I'm missing something). Anyway this commit doesn't need to be kept as removing the whole method also works.

**Comment by nalimilan** [comment]
@tkelman Trying to find the simplest fix which would avoid data loss until Pkg3 lands, what kind of solution to do you suggest?

**Comment by tkelman** [comment]
From last time I looked at this I got to about the same place you did. I think the corruption is due to libgit2 or the windows filesystem not being completely done with the checkout operation before calling the resolve

**Comment by nalimilan** [comment]
So do you think that the PR without the last commit is correct? It doesn't fix #17994 at all, but at least that would turn this question into a design choice (remove or not) rather than into a mix of bug+design.

**Comment by tkelman** [comment]
It would avoid the corruption at least, but it would also change the error recovery behavior in a way that I don't think is ideal. Is it possible to use 2 nested levels of LibGit2.transact to maybe get both without changing overall recovery behavior if resolve errors?

**Comment by nalimilan** [comment]
Honestly I don't understand the subtleties of that code, but AFAICT we cannot call `resolve` from within `LibGit2.transact` since it keeps file handles open. Maybe we could set a boolean flag in case of success, and only call `resolve` if it's set? BTW, what's the problem exactly with calling `resolve` after a git failure? I guess any other Pkg operation may call `resolve` later anyway, so if that's a problem it will bite at some later point, right?

**Comment by tkelman** [comment]
If something wrong with the git operation causes a failure, it should attempt to roll back to the previous state instead of leaving things in whatever in between condition they were at when the error happened.

**Comment by nalimilan** [comment]
> If something wrong with the git operation causes a failure, it should attempt to roll back to the previous state instead of leaving things in whatever in between condition they were at when the error happened.

But the current code doesn't do that, right? I really don't want to dive into complex improvements of the Pkg system since it's going to be improved, I'm just trying to find the smallest backportable fix to avoid data loss.

**Comment by tkelman** [comment]
https://github.com/JuliaLang/julia/blob/19b3ca7be9f47099dcf192b57e3b044c571ed155/base/libgit2/libgit2.jl#L892-L899
that's the intent of transact?

**Comment by nalimilan** [comment]
OK, I hadn't noticed the `pkgs` method didn't use `transact`. That's kind of weird. Should I just add it?

**Comment by nalimilan** [comment]
Not going to happen at this point.

**Pull request closed without merging**